### PR TITLE
[spirv] Add lowering config verifier for CooperativeMatrix pipeline

### PR DESCRIPTION
**Pull request by yzhang93**

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 1]
```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
...
+    return success();
+  }
+  LLVM_DEBUG({
+    llvm::dbgs() << "verifying op: " << *op << "\n";
+    llvm::dbgs() << "chosen workgroup size: [";
+    llvm::interleaveComma(workgroupSize, llvm::dbgs());
+    llvm::dbgs() << "]\n";
+  });
+
+  // Get spirv.target_env attributes
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:176]: Nit: place `.` to end the sense for the comment. Also for the following comments.

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 2]
```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
...
+  // We don't check the batch.
+  if (isa<linalg::BatchMatmulOp>(op)) {
+    lhsShape = lhsShape.drop_front(1);
+    rhsShape = rhsShape.drop_front(1);
+    firstLevelTileSizes.erase(firstLevelTileSizes.begin());
+    secondLevelTileSizes.erase(secondLevelTileSizes.begin());
+    thirdLevelTileSizes.erase(thirdLevelTileSizes.begin());
+  }
+
+  // Verify that subgroup tile sizes should be multiple of 16
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:257]: Actually `16x16x16` is just one configuration. We should first try to search the cooperative matrix property list in resource limits to find a match for the 4th level (ignoring the leading 1 if batch matmul). Then use the found match as the native matmul size.

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 3]
```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
...
+           << 2 * subgroupSize;
+  }
+
+  Type inputType = op->getOperand(0).getType();
+  ArrayRef<int64_t> lhsShape =
+      op->getOperand(0).getType().cast<ShapedType>().getShape();
+  ArrayRef<int64_t> rhsShape =
+      op->getOperand(1).getType().cast<ShapedType>().getShape();
+
+  SmallVector<int64_t> firstLevelTileSizes =
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:235]: s/firstLevel/workgrup/?

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 4]
```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
...
+      loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+  SmallVector<int64_t> secondLevelTileSizes =
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:237]: s/secondLevel/subgroup/?

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 5]
```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
...
+      loweringConfig.getTileSizeVals(kThreadTileLevel);
+  SmallVector<int64_t> thirdLevelTileSizes =
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:239]: s/thirdLevel/reduction/?

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 6]
```diff
@@ -150,6 +150,155 @@ LogicalResult verifySPIRVMatmulPromoteVectorizePassPipeline(
...
+    secondLevelTileSizes.erase(secondLevelTileSizes.begin());
+    thirdLevelTileSizes.erase(thirdLevelTileSizes.begin());
+  }
+
+  // Verify that subgroup tile sizes should be multiple of 16
+  if (secondLevelTileSizes[0] % 16 != 0 || secondLevelTileSizes[1] % 16 != 0 ||
+      thirdLevelTileSizes[2] % 16 != 0) {
+    return op->emitOpError("expected subgroup tile sizes to be multiple of 16");
+  }
+
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:262]: Define `const int64_t dimM = lhsShape[0], dimN = rhsShape[1], dimK = lhsShape[1];` to make it more readable?

**Review by antiagainst — CHANGES_REQUESTED** [review]
Nice, thanks! Just a few issues.

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp (hunk 3 above)
Comment (yzhang93) [reply compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:235]: Do you mean to change the variable name from firstLevelTileSizes to something like workgroupTileSizes?

Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:235]: Yup!

File: compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp [hunk 7]
```diff
@@ -249,45 +247,55 @@ LogicalResult verifySPIRVCooperativeMatrixVectorizePassPipeline(
...
-
-  // Verify the tile size divides the matmul inputs A [M x K] & B [K x N]
-  if (lhsShape[0] % firstLevelTileSizes[0] != 0 ||
-      lhsShape[1] % thirdLevelTileSizes[2] != 0) {
+    workgroupTileSizes.erase(workgroupTileSizes.begin());
+    subgroupTileSizes.erase(subgroupTileSizes.begin());
+    reductionTileSizes.erase(reductionTileSizes.begin());
+  }
+
+  // Verify that subgroup tile sizes should be multiple of cooperative matrix
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp:255]: Actually as said before, we should first check the fourth level against the cooperative matrix properties. Something like the following:

```
SmallVector<int64_t> nativeVectorSizes = loweringConfig.getTileSizeVals(3);
...
auto getElementType = [](Value v) { return v.getType().cast<ShapedType>().getElementType(); };
Type aType = getElementType(lhs);
Type bType = getElementType(rhs);
Type cType = getElementType(result);
...
if (isa<linalg::BatchMatmulOp>(op)) {
  ...
  nativeVectorSizes.erase(nativeVectorSizes.begin());
}
...

bool isNativeVectorSizeAccepted = false;
for (auto p : properties) {
  if (p.getAType() == aType && p.getBType() == bType && p.getCType == cType() &&
      p.getScope().getValue() == spirv::Scope::Subgroup &&
      p.getMSize() == nativeVectorSizes[0] && p.getNSize() == nativeVectorSizes[1] && p.getKSize() == nativeVectorSizes[2])
  isNativeVectorSizeAccepted = true;
}
if (!isNativeVectorSizeAccepted) op->emitOpError(...);

// Then go on to check subgroup tile sizes is a multiple of nativeVectorSizes
```

Comment (yzhang93) [reply compiler/src/iree/compiler/Codegen/SPIRV/Verifiers.cpp]: Thanks for pointing out the problem. I've made the changes.

**Review by antiagainst — APPROVED** [review]

**Comment by antiagainst** [comment]
Nice, thanks for being patient and fixing all the issues! :D

**Pull request merged**